controllers/postgresqldatabase_controller.go

```diff
@@ -223,14 +263,90 @@ func (r *PostgreSQLDatabaseReconciler) EnsurePostgreSQLDatabase(log logr.Logger,
 			log.Error(err, "failed to close database connection", "host", host, "database", "postgres", "user", credentials.Name)
 		}
 	}()
-	err = postgres.Database(log, db, host, postgres.Credentials{
-		Name:     name,
-		User:     user,
-		Password: password,
-		Shared:   isShared,
-	})
+	err = postgres.Database(log, db, host, params.Target)
 	if err != nil {
-		return fmt.Errorf("create database %s on host %s: %w", name, connectionString, err)
+		return fmt.Errorf("create database %s on host %s: %w", params.Target.Name, connectionString, err)
 	}
 	return nil
 }
+
+// credentials returns the correct `postgres.Credentials` based on the values
+// of `params.Host` and `params.HostCredentials` fields of `params`. Exactly
+// one of these fields must be set, and if `Host` is set, then this method
+// will return the credentials from `r.HostCredentials` map, otherwise it will
+// search the Kubernetes namespace for a `PostgreSQLHostCredentials` with the
+// name specified in `params.HostCredentials`.
+func (r *PostgreSQLDatabaseReconciler) credentials(ctx context.Context, params *EnsureParams) (string, *postgres.Credentials, error) {
+	// If the `HostCredentials` field is populated and the `Host` field is
+	// empty, then return credentials stored in the corresponding
+	// `PostgreSQLHostCredentials` resource.
+	if params.Host == "" && params.HostCredentials != "" {
+		// Fetch the `PostgreSQLHostCredentials` from the API.
+		var hostCreds postgresqlv1alpha1.PostgreSQLHostCredentials
+		if err := r.Client.Get(
+			ctx,
+			types.NamespacedName{
+				Namespace: params.Namespace,
+				Name:      params.HostCredentials,
+			},
+			&hostCreds,
+		); err != nil {
+			if apierrors.IsNotFound(err) {
+				return "", nil, &ctlerrors.Invalid{
+					Err: fmt.Errorf("unknown credentials for host"),
+				}
```

Comment:
You can use `ctlerrors.NewInvalid()`.

```suggestion
				return "", nil, ctlerrors.NewInvalid(fmt.Errorf("unknown credentials for host"))
```